src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,421 @@
+        public static List<string> GetSerialPortNames()
+        {
+            return SerialPort.GetPortNames().ToList();
+        }
+
+        /// <summary>
+        /// Returns a list of commonly used baud rates.
+        /// </summary>
+        public static List<int> CommonBaudRates()
+        {
+            return new List<int>()
+            {
+                9600,
+                19200,
+                18400,
+                57600,
+                115200,
+                230400,
+                250000,
+                500000,
+                1000000,
+                2000000,
+            };
+        }
+
+        /// <summary>
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        public virtual void Initialize()
+        {
+            if (_firmata != null)
+            {
+                throw new InvalidOperationException("Board already initialized");
+            }
+
+            _firmata = new FirmataDevice();
+            _firmata.Open(_dataStream);
+            _firmata.OnError += FirmataOnError;
+            _protocolVersion = _firmata.QueryFirmataVersion();
+            if (_protocolVersion < _firmata.QuerySupportedFirmataVersion())
+            {
+                throw new NotSupportedException($"Firmata version on board is {_protocolVersion}. Expected {_firmata.QuerySupportedFirmataVersion()}. They must be equal.");
+            }
+
+            Log($"Firmata version on board is {_protocolVersion}.");
+
+            _firmwareVersion = _firmata.QueryFirmwareVersion(out _firmwareName);
+
+            Log($"Firmware version on board is {_firmwareVersion}");
+
+            _firmata.QueryCapabilities();
+
+            _supportedPinConfigurations = _firmata.PinConfigurations; // Clone reference
+
+            Log("Device capabilities: ");
+            foreach (var pin in _supportedPinConfigurations)
+            {
+                Log(pin.ToString());
+            }
+
+            _firmata.EnableDigitalReporting();
+        }
+
+        /// <summary>
+        /// Firmware version on the device
+        /// </summary>
+        public Version FirmwareVersion
+        {
+            get
+            {
+                return _firmwareVersion ?? new Version();
+            }
+        }
+
+        /// <summary>
+        /// Name of the firmware.
+        /// </summary>
+        public string FirmwareName
+        {
+            get
+            {
+                return _firmwareName ?? string.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Firmata version found on the board.
+        /// </summary>
+        public Version FirmataVersion
+        {
+            get
+            {
+                if (_protocolVersion == null)
+                {
+                    return new Version();
```

Comment:
should we throw when we don't know protocol version?

Reply:
That happens in the Initialize function.